bittide-instances/exe/post-vex-riscv-tcp-test/Main.hs

```diff
@@ -80,51 +100,78 @@ case_testGdbProgram = do
-        hFlush stdout
-        waitForLine picocomStdOutHandle "Terminal ready"
-        putStrLn " Done"
+        putStrLn "Waiting for Picocom to be ready..."
+        _ <-
+          tryWithTimeout "Timeout waiting for Picocom to be ready" 10_000_000 $
+            waitForLine picocomStdOutHandle "Terminal ready"
 
         putStrLn "Starting GDB..."
         withCreateProcess gdbProc $ \_ (fromJust -> gdbStdOut) _ _ -> do
-          -- Wait for GDB to program the FPGA. If successful, we should see
-          -- "going in echo mode" in the picocom output.
           hSetBuffering gdbStdOut LineBuffering
 
-          putStr "Waiting for GDB to program the FPGA..."
-          hFlush stdout
-          waitForLine picocomStdOutHandle "listening"
-          putStrLn " Done"
-
-        putStrLn "Starting TCP Spray..."
-        -- Test TCP echo server
-        withCreateProcess tcpSprayProc $ \_ (fromJust -> tcpStdOut) (fromJust -> tcpStdErr) _ -> do
-          (stdOutContents, stdErrContents) <- do
-            stdOutContents' <- hGetContents' tcpStdOut
-            stdErrContents' <- hGetContents' tcpStdErr
-            pure (stdOutContents', stdErrContents')
-          picocomOut <- hGetContents picocomStdOutHandle
-
-          case maybePicocomStdErr of
-            Nothing -> pure ()
-            Just h -> do
-              putStrLn "Picocom StdErr"
-              readRemainingChars h >>= putStrLn
-
-          putStrLn ""
-          putStrLn "Picocom stdout"
-          putStrLn $ readUntil "listening" picocomOut
-
-          putStrLn ""
-          putStrLn "tcpSpray StdOut"
-          putStrLn stdOutContents
-
-          if not (null stdErrContents)
-            then assertFailure $ "tcpspray failed with stderr:\n" <> stdErrContents
-            else pure ()
+          putStrLn "Waiting for \"Starting TCP Client\""
+          _ <-
+            tryWithTimeout "Timeout waiting for \"Starting TCP Client\"" 10_000_000 $
+              waitForLine picocomStdOutHandle "Starting TCP Client"
+
+          let numberOfClients = 1
+          putStrLn $ "Waiting for " <> show numberOfClients <> " clients to connect to TCP server."
+          clients <-
+            tryWithTimeout "Timeout waiting for clients to connect" 10_000_000 $
+              waitForClients numberOfClients serverSock
+
+          putStrLn "Writing client data to files"
+          createDirectoryIfMissing True tcpDataDir
+          mapConcurrently_ runTcpTest clients
+          putStrLn "Closing client connections"
+          mapConcurrently_ (NS.closeSock . fst) clients
+          loggingSequence
+
+runTcpTest :: (NS.Socket, NS.SockAddr) -> Assertion
+runTcpTest (sock, sockAddr) = do
+  putStrLn $ "Testing TCP connection from " <> show sockAddr
+  start <- getPOSIXTime
+  bs <- SBS.toLazy_ $ SBS.reread (const $ NS.recv sock (1024 * 1024)) ()
+  stop <- getPOSIXTime
+  let size = BS.length bs
+  let speed = fromIntegral size / (stop - start)
```

Comment:
You need some extra type conversion here.
Because now `speed :: NominalDiffTime`, and it will be shown with a "s" suffix, because it is supposed to represent seconds.